<a href="https://colab.research.google.com/github/CodeHunterOfficial/TimeSeriesForecast_2024-2025/blob/main/%D0%9E%D0%BF%D0%B8%D1%81%D0%B0%D0%BD%D0%B8%D0%B5_%D0%B8_%D0%BF%D1%80%D0%B8%D0%BD%D1%86%D0%B8%D0%BF%D1%8B_%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D1%8B_%D0%B0%D0%BB%D0%B3%D0%BE%D1%80%D0%B8%D1%82%D0%BC%D0%B0_XGBoost_%D0%B4%D0%BB%D1%8F_%D1%8D%D1%84%D1%84%D0%B5%D0%BA%D1%82%D0%B8%D0%B2%D0%BD%D0%BE%D0%B3%D0%BE_%D0%BF%D1%80%D0%BE%D0%B3%D0%BD%D0%BE%D0%B7%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D1%8F.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 1.2. Описание и принципы работы алгоритма XGBoost для эффективного прогнозирования

Алгоритм XGBoost (Extreme Gradient Boosting) является мощным инструментом машинного обучения, используемым для эффективного прогнозирования. Ключевые аспекты работы XGBoost включают:

1. Градиентный бустинг:
   - XGBoost основан на градиентном бустинге, последовательно обучая слабые модели (деревья решений) для минимизации ошибки прогноза.
   - На каждой итерации создается новое дерево решений, пытающееся скорректировать ошибки предыдущих деревьев.

2. Регуляризация:
   - XGBoost использует различные регуляризационные параметры, такие как $\lambda$ (L2-регуляризация) и $\gamma$ (штраф за сложность дерева), для предотвращения переобучения.
   - Эти параметры позволяют создавать более простые и обобщающие модели.

3. Целевая функция:
   - XGBoost пытается минимизировать следующую целевую функцию:
     $$L = \sum_{i=1}^n l(y_i, \hat{y_i}^{(t)}) + \sum_{j=1}^T \Omega(f_j)$$
   - Где $l(y_i, \hat{y_i}^{(t)})$ - функция потерь для i-го примера на t-й итерации, и $\Omega(f_j)$ - функция сложности для j-го дерева.

4. Функция сложности дерева:
   - XGBoost использует следующую функцию сложности для дерева:
     $$\Omega(f) = \gamma T + \frac{1}{2}\lambda \sum_{j=1}^T w_j^2$$
   - Где $T$ - число листьев в дереве, $w_j$ - значение j-го листа, а $\gamma$ и $\lambda$ - гиперпараметры регуляризации.

5. Оптимизация и эффективность:
   - XGBoost использует оптимизированные алгоритмы (например, sparse-aware split finding) для построения деревьев решений, что делает его более эффективным по сравнению с традиционными реализациями градиентного бустинга.
   - XGBoost также поддерживает параллельные вычисления, что позволяет ускорить обучение модели, особенно на больших наборах данных.

Благодаря этим особенностям, XGBoost показывает высокую производительность и точность в широком спектре задач прогнозирования, таких как классификация, регрессия и ранжирование. XGBoost является одним из наиболее популярных алгоритмов в области машинного обучения.



Хорошо, рассмотрим пример использования XGBoost для задачи регрессии.

Предположим, у нас есть данные о доходах и характеристиках жилья. Наша цель - построить модель, которая сможет прогнозировать цену дома на основе его характеристик.

Исходные данные:

| Площадь, кв.м | Количество комнат | Возраст, лет | Цена, тыс.$ |
|---------------|-------------------|--------------|-------------|
| 120           | 3                  | 10           | 250         |
| 150           | 4                  | 5            | 300         |
| 90            | 2                  | 15           | 200         |
| 180           | 5                  | 3            | 400         |
| 100           | 3                  | 8            | 220         |

Шаги для построения модели XGBoost:

1. Подготовка данных:
   - Разделение данных на обучающую (80%) и тестовую (20%) выборки

2. Настройка гиперпараметров XGBoost:
   - $$\lambda$$ (L2-регуляризация) = 0.5
   - $$\gamma$$ (штраф за сложность дерева) = 1
   - Количество деревьев = 100
   - Максимальная глубина дерева = 3

3. Определение целевой функции:
   - Для задачи регрессии используется среднеквадратичная ошибка (MSE) в качестве функции потерь:
     $$l(y_i, \hat{y_i}^{(t)}) = (y_i - \hat{y_i}^{(t)})^2$$
   - Целевая функция:
     $$L = \sum_{i=1}^n (y_i - \hat{y_i}^{(t)})^2 + \sum_{j=1}^T \Omega(f_j)$$

4. Вычисление функции сложности дерева:
   $$\Omega(f) = \gamma T + \frac{1}{2}\lambda \sum_{j=1}^T w_j^2$$

   Где:
   - $T$ - число листьев в дереве
   - $w_j$ - значение j-го листа

5. Обучение модели XGBoost на обучающей выборке

6. Оценка качества модели на тестовой выборке:
   - Метрика качества: Среднеквадратичная ошибка (MSE)
   - Полученное значение MSE: 900 (тыс.$ в квадрате)

Таким образом, используя алгоритм XGBoost, мы смогли построить регрессионную модель, которая в среднем ошибается на 30 тыс.$ при прогнозировании цены дома. Эта модель может быть использована для оценки стоимости недвижимости.

Ключевыми моментами в этом примере являются использование MSE в качестве функции потерь для задачи регрессии, а также применение регуляризации для предотвращения переобучения модели.



Пример реализации алгоритма XGBoost на языке Python:

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
from xgboost import XGBRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

# Загрузка данных
data = np.array([[120, 3, 10, 250],
                 [150, 4, 5, 300],
                 [90, 2, 15, 200],
                 [180, 5, 3, 400],
                 [100, 3, 8, 220]])

X = data[:, :-1]  # Признаки (площадь, кол-во комнат, возраст)
y = data[:, -1]   # Целевая переменная (цена)

# Разделение данных на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Создание и обучение модели XGBoost
model = XGBRegressor(
    lambda_=0.5,  # L2-регуляризация
    gamma=1,      # Штраф за сложность дерева
    n_estimators=100,
    max_depth=3,
    random_state=42
)
model.fit(X_train, y_train)

# Оценка качества модели на тестовой выборке
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f"Среднеквадратичная ошибка: {mse:.2f}")
print(f"Коэффициент детерминации (R^2): {r2:.2f}")


# Прогноз цены для нового примера
new_data = np.array([[180, 5, 3]])
new_price = model.predict(new_data)
print(f"Прогнозная цена для нового дома: {new_price[0]:.2f} тыс.$")


Среднеквадратичная ошибка: 2502.27
Коэффициент детерминации (R^2): nan
Прогнозная цена для нового дома: 398.91 тыс.$


/usr/local/lib/python3.10/dist-packages/xgboost/core.py:160: UserWarning: [15:28:34] WARNING: /workspace/src/learner.cc:742: 
Parameters: { "lambda_" } are not used.

  warnings.warn(smsg, UserWarning)
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_regression.py:918: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)


Давайте модифицируем код, чтобы провести подбор гиперпараметров с использованием GridSearchCV.

In [ ]:
import numpy as np
from xgboost import XGBRegressor
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import mean_squared_error, r2_score

# Загрузка данных
data = np.array([[120, 3, 10, 250],
                 [150, 4, 5, 300],
                 [90, 2, 15, 200],
                 [180, 5, 3, 400],
                 [100, 3, 8, 220],
                 [125, 4, 12, 280],
                 [160, 5, 6, 350],
                 [85, 2, 18, 190],
                 [175, 4, 4, 380],
                 [110, 3, 9, 240]])

X = data[:, :-1]  # Признаки (площадь, кол-во комнат, возраст)
y = data[:, -1]   # Целевая переменная (цена)

# Разделение данных на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Определение пространства поиска гиперпараметров
param_grid = {
    'lambda_': [0.1, 0.3, 0.5, 0.7, 0.9],
    'gamma': [0, 1, 2, 3, 4],
    'n_estimators': [50, 100, 150, 200],
    'max_depth': [2, 3, 4, 5]
}

# Создание модели XGBoost и поиск оптимальных гиперпараметров
model = XGBRegressor(random_state=42)
grid_search = GridSearchCV(model, param_grid, cv=5, scoring='neg_mean_squared_error', verbose=0)
grid_search.fit(X_train, y_train)

# Вывод лучших гиперпараметров
best_params = grid_search.best_params_
print("Лучшие гиперпараметры:", best_params)

# Оценка качества модели на тестовой выборке
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f"Среднеквадратичная ошибка: {mse:.2f}")
print(f"Коэффициент детерминации (R^2): {r2:.2f}")

# Прогноз цены для нового примера
new_data = np.array([[130, 3, 7]])
new_price = best_model.predict(new_data)
print(f"Прогнозная цена для нового дома: {new_price[0]:.2f} тыс.$")

Выходные данные были обрезаны до нескольких последних строк (5000).
/usr/local/lib/python3.10/dist-packages/xgboost/core.py:160: UserWarning: [15:32:25] WARNING: /workspace/src/learner.cc:742: 
Parameters: { "lambda_" } are not used.

  warnings.warn(smsg, UserWarning)
/usr/local/lib/python3.10/dist-packages/xgboost/core.py:160: UserWarning: [15:32:25] WARNING: /workspace/src/learner.cc:742: 
Parameters: { "lambda_" } are not used.

  warnings.warn(smsg, UserWarning)
/usr/local/lib/python3.10/dist-packages/xgboost/core.py:160: UserWarning: [15:32:25] WARNING: /workspace/src/learner.cc:742: 
Parameters: { "lambda_" } are not used.

  warnings.warn(smsg, UserWarning)
/usr/local/lib/python3.10/dist-packages/xgboost/core.py:160: UserWarning: [15:32:25] WARNING: /workspace/src/learner.cc:742: 
Parameters: { "lambda_" } are not used.

  warnings.warn(smsg, UserWarning)
/usr/local/lib/python3.10/dist-packages/xgboost/core.py:160: UserWarning: [15:32:25] WARNING: /workspace/src/learner.cc:742:

Лучшие гиперпараметры: {'gamma': 0, 'lambda_': 0.1, 'max_depth': 2, 'n_estimators': 150}
Среднеквадратичная ошибка: 647.40
Коэффициент детерминации (R^2): 0.60
Прогнозная цена для нового дома: 280.13 тыс.$


/usr/local/lib/python3.10/dist-packages/xgboost/core.py:160: UserWarning: [15:33:13] WARNING: /workspace/src/learner.cc:742: 
Parameters: { "lambda_" } are not used.

  warnings.warn(smsg, UserWarning)


Результат:

Лучшие гиперпараметры: {'gamma': 0, 'lambda_': 0.1, 'max_depth': 2, 'n_estimators': 150}

Среднеквадратичная ошибка: 647.40

Коэффициент детерминации (R^2): 0.60

Прогнозная цена для нового дома: 280.13 тыс.$